### PR TITLE
Add Nixpkgs support for Dhall

### PR DESCRIPTION
**Pull request by Gabriella439**
One of the motivations for this change is the following Discourse
discussion:

https://discourse.dhall-lang.org/t/offline-use-of-prelude/137

Many users have requested Dhall support for "offline" packages
that can be fetched/built/installed using ordinary package
management tools (like Nix) instead of using Dhall's HTTP import system.
I will continue to use the term "offline" to mean Dhall package
builds that do not use Dhall's language support for HTTP imports (and
instead use the package manager's support for HTTP requests, such
as `pkgs.fetchFromGitHub`)

The goal of this change is to document what is the idiomatic way to
implement "offline" Dhall builds by implementing Nixpkgs support
for such builds.  That way when other package management tools ask
me how to package Dhall with their tools I can refer them to how it
is done in Nixpkgs.

This change contains a fully "offline" build for the largest Dhall
package in existence, known as "dhall-packages" (not to be confused
with `dhallPackages`, which is our Nix attribute set containing
Dhall packages).

The trick to implementing offline builds in Dhall is to take
advantage of Dhall's support for semantic integrity checks.  If an
HTTP import is protected by an integrity check and a cached build
product matches the integrity check then the HTTP import is never
resolved and the expression is instead fetched from cache.

By "installing" dependencies in a pre-seeded and isolated cache
we can replace remote HTTP imports with dependencies that have
been built and supplied by Nix instead.

The offline nature of the builds are enforced by compiling the
Haskell interpreter with the `-f-with-http` flag, which disables
the interpreter's support for HTTP imports.  If a user forgets
to supply a necessary dependency as a Nix build product then the
build fails informing them that HTTP imports are disabled.

By default, built packages are "binary distributions", containing
just a cache product and a Dhall expression which can be used to
resolve the corresponding cache product.

Users can also optionally enable a "source distribution" of a package
which already includes the equivalent fully-evaluated Dhall code (for
convenience), but this is disabled by default to keep `/nix/store`
utilization as compact as possible.

### CODE REVIEW

File: pkgs/development/dhall-modules/Prelude.nix
Comment (Mic92) [inline pkgs/development/dhall-modules/Prelude.nix]: Do we really need all those versions?

Comment (Gabriel439) [reply pkgs/development/dhall-modules/Prelude.nix]: @Mic92: Not really.  I just added as many as possible because it was easy to do.  I can remove them if desired

**Comment by Profpatsch** [comment]
Will review tomorrow.

File: pkgs/top-level/dhall-packages.nix
Comment (Profpatsch) [inline pkgs/top-level/dhall-packages.nix]: what’s the reasoning behind this attribute? It doesn’t seem to contain anything that the others don’t expose.

File: pkgs/top-level/dhall-packages.nix
Comment (Profpatsch) [inline pkgs/top-level/dhall-packages.nix]: Not a fan of the complexity introduced by this scoping override (not a fan of how `haskellPackages` overcomplicates many things in general). Can we keep it simple? `let` is a great idea.

**Review by Profpatsch — CHANGES_REQUESTED** [review]
I haven’t done a full review, but it looks to me like this was mostly inspired by the `haskellPackages` code, which is a horribly clunky thing in my opinion. I’d advise to keep it more simplistic.

File: pkgs/top-level/dhall-packages.nix
Comment (Gabriel439) [reply pkgs/top-level/dhall-packages.nix]: @Profpatsch: Actually, I was following how `pythonPackages` did things (which happens to be the same as `haskellPackages`)

I'm not exactly sure what alternative approach you are suggesting, but it sounds like you are recommending that I use `pkgs.callPackages` (with no scoping override).  I gave that a try and ran into the following issues:

* It's now more difficult for one Dhall package to depend upon another Dhall package

  ... since the non-overriden `callPackage` doesn't bring any Dhall packages in scope

* You lose the ability to override `buildDhallPackage`

  ... which is really useful, for the same reason that it's useful to override `pkgs.haskellPackage.mkDerivation`.  For example, you could globally override `buildDhallPackage` to set `source = true;` to build source distributions for Dhall packages

Or did I misunderstand what you are requesting?  Perhaps if you included a short code sample of what you had in mind it would be more clear

File: pkgs/top-level/dhall-packages.nix
Comment (Gabriel439) [reply pkgs/top-level/dhall-packages.nix]: `dhall-packages` does not just re-export other Dhall packages.  `dhall-packages` provides new functionality of its own (like Argo CD bindings, for example)

File: pkgs/top-level/dhall-packages.nix
Comment (Profpatsch) [reply pkgs/top-level/dhall-packages.nix]: I’m sorry, I was in a bad mood yesterday and did this review against my better judgement. I should do better than that.

We should use the approach in this PR because of two reasons:

1) It’s the default, with precedence set by `haskellPackages` and `pythonPackages`. There’s a lot of value in regularity, even if the usability or design is not the greatest.
2) There hasn’t been a better approach so far.

File: pkgs/top-level/dhall-packages.nix
Comment (Profpatsch) [reply pkgs/top-level/dhall-packages.nix]: Oh, I see, `dhall-packages` is a (mono)repository containing other dhall packages.

File: pkgs/development/dhall-modules/dhall-packages.nix
Comment (Profpatsch) [inline pkgs/development/dhall-modules/dhall-packages.nix]: So if `dhall-packages` (confusing name!) is a monorepository providing vendored versions of other packages, this marks a shortcoming of the proposed approach; how would I as a nix maintainer keep these definitions up to date?
Do I have to rummage through the repository to find all dependencies and manually resolve their version to patch them here? Can dhall help me with this?

Comment (Gabriel439) [reply pkgs/development/dhall-modules/dhall-packages.nix]: @Profpatsch: My plan was to follow this up with a `dhall-to-nixpkgs` utility, which would play a role similar to other `*2nix` utilities like `cabal2nix`.  You would give it a source directory or a URL as input and it would translate that to a derivation matching the style as those underneath `./pkgs/development/dhall-modules`

Essentially the tool would:

* Guess the name of the Dhall package from the directory name or URL
* Obtain all external dependencies by scanning transitive relative imports
* Turn those external dependencies into the `dependencies` field of `buildDhallPackage` (using the same rule for inferring package names)

For example, if it were to see an external dependency of the form:

```dhall
https://raw.githubusercontent.com/${user}/${repo}/${version}/${path}
```

Then it would add this to the `dependencies` list:

```dhall
  (repo."${version}".override { file = path })
```

... or we could include the `user` in the package name, too, in order to minimize conflicts.

Then the next logical step would be writing a `callDhall2Nixpkgs` analogous to `callCabal2nix`

**Review by Profpatsch — APPROVED** [review]
Before I block this any further, let’s merge.

I’m sure you are going to come up with a satisfactory solution iteratively, like you always seem to do :+1:

**Comment by ehmry** [comment]
How does this work without any setup or shell hooks? Why merge if it isn't used anywhere in the tree?

**Pull request merged**
